swarm/network/pss.go

```diff
@@ -1,52 +1,384 @@
 package network
 
 import (
-	"fmt"
 	"bytes"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"time"
 
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/logger"
 	"github.com/ethereum/go-ethereum/logger/glog"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/adapters"
+	"github.com/ethereum/go-ethereum/p2p/protocols"
+	"github.com/ethereum/go-ethereum/pot"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+const (
+	DefaultTTL                  = 6000
+	TopicLength                 = 32
+	TopicResolverLength         = 8
+	PssPeerCapacity             = 256
+	PssPeerTopicDefaultCapacity = 8
+	digestLength                = 64
+	digestCapacity              = 256
+	defaultDigestCacheTTL       = time.Second
 )
 
+var (
+	zeroRW    = PssReadWriter{}
+	zeroTopic = PssTopic{}
+)
+
+type PssParams struct {
+}
+
+func NewPssParams() *PssParams {
+	return &PssParams{}
+}
+
+type PssMsg struct {
+	To      []byte
+	Payload PssEnvelope
+}
+
+func (pm *PssMsg) String() string {
+	return fmt.Sprintf("PssMsg: Recipient: %x", pm.To[:8])
+}
+
+// Pre-Whisper placeholder
+type PssEnvelope struct {
+	Topic       PssTopic
+	TTL         uint16
+	Payload     []byte
+	SenderOAddr []byte
+	SenderUAddr []byte
+}
+
+// Pre-Whisper placeholder
+type pssPayload struct {
+	Code       uint64
+	Size       uint32
+	Data       []byte
+	ReceivedAt time.Time
+}
+
+type PssTopic [TopicLength]byte
+
+type pssDigest uint32
+
+// Pss master object provides:
+// - access to the swarm overlay and routing (kademlia)
+// - a collection of remote overlay addresses mapped to MsgReadWriters, representing the virtually connected peers
+// - a collection of remote underlay address, mapped to the overlay addresses above
+// - a method to send a message to specific overlayaddr
 type Pss struct {
-	Overlay
-	LocalAddr	[]byte
-	C	chan []byte
+	Overlay                                              // we can get the overlayaddress from this
+	NodeId   *adapters.NodeId                            // we need the underlayaddr to make virtual p2p.Peers in the receiving end
+	PeerPool map[pot.Address]map[PssTopic]*PssReadWriter // keep track of all virtual p2p.Peers we are currently speaking to
+	//PeerReverse map[adapters.NodeId]pot.Address // keep track of all virtual p2p.Peers we are currently speaking to
+	handlers map[PssTopic]func([]byte, *p2p.Peer, []byte) error // topic and version based pss payload handlers
+	fwdcache map[pssDigest]time.Time                            // checksum of unique fields from pssmsg mapped to expiry, cache to determine whether to drop msg
+	cachettl time.Duration                                      // how long to keep messages in fwdcache
+	hasher   hash.Hash                                          // hasher to digest message to cache
 }
 
-func NewPss(k Overlay, addr []byte) *Pss {
+// 32bit checks should be enough for this use
+func (self *Pss) hashMsg(msg *PssMsg) pssDigest {
```

Comment:
i'm afraid, this will be called asynchronously, so you need to use
`hasher func() hash.Hash` on line 87